Docs: features (#106)
chakra-ui/chakra-ui-vue-next: docs/pages/features/chakra-factory.md

# Chakra Factory

> Chakra factory serves as an **object of chakra enabled JSX elements**, and also

Reviewer: ```suggestion
Chakra factory serves as an **object of chakra enabled HTML elements**, and also
```

**a function that can be used to enable custom component** to receive chakra's
style props.

```bash
import { chakra } from "@chakra-ui/vue-next"
```

## Chakra JSX Elements

Create base HTML elements with theme-aware style props using `chakra.<element>`
notation. For example, if you want a plain HTML button with the ability to pass
chakra styles, you can write `<chakra.button />`.

> ```jsx
> <chakra.button
>   px="3"
>   py="2"
>   bg="green.200"
>   rounded="md"
>   :_hover="{ bg: "green.300" }"
> >
>   Click me
> </chakra.button>
> ```

Reviewer: I think we can directly use Vue SFCs for the docs codeblocks

```suggestion
```vue
<chakra.button
  px="3"
  py="2"
  bg="green.200"
  rounded="md"
  :_hover="{ bg: 'green.300' }"
>
  Click me
</chakra.button>
```
```

This reduces the need to create custom component wrappers and name them. This
syntax is available for common HTML elements. See the reference for the full
[list of elements](https://github.com/chakra-ui/chakra-ui/blob/main/packages/system/src/system.utils.ts#L9)
supported.

```jsx
<chakra.h1 fontSize="lg"> Heading </chakra.h1>
```

## Chakra factory function

This is a function that converts **non-chakra components** or **jsx element** to
chakra-enabled components so you can pass style props to them.

> Consider a package called `my-awesome-components`, let's use the chakra factory

Reviewer: ```suggestion
Consider a component called `my-awesome-component`, let's use the chakra factory
```

function to make it possible to pass style props.

The function will infer the prop types from the wrapped component and also add
chakra style props.

```vue
<template>
  <ChakraAwesomeComponent bg="red.200" fontSize="12px" />
</template>

<script setup>
  import { chakra } from "@chakra-ui/vue-next"
  import awesomeComponent from "my-awesome-components"

  const ChakraAwesomeComponent = chakra(awesomeComponent)
</script>
```

> Considering that Chakra uses `emotion` under the hood, ensure the non-chakra
> component accepts `className` as props for this to work correctly

### Attaching styles

In some instances, you might need to attach specific styles to the component
wrapped in the chakra factory

```jsx
const ChakraAwesomeComponent = chakra(awesomeComponent, {
  baseStyle: {
    bg: "papayawhip",
    color: "red.500"
  },
})
```

> You can also use the chakra factory on jsx-elements as well.

Reviewer: ```suggestion
You can also use the chakra factory on HTML elements as well.
```
